### PR TITLE
7770 facebook crosspost with links

### PR DESCRIPTION
**Pull request by HankG**
A review of submission against https://github.com/diaspora/diaspora/issues/7770

Spec has been updated for new cases, style check has been run locally, but I don't know what rules for integration testing are before integration into main line since I don't have ability to test Facebook integration directly. If this looks good and the spec updates alone are sufficient then this may be ready for merge rather than just review.  On the style check some code I didn't write wasn't passing and I couldn't figure out how to consistently get around the "Align the elements of a hash literal if they span more than one line" warning no matter what permutations I tried so I made everything that flagged it single line.  I'll look for guidance on that on Discourse channels...

### CODE REVIEW

File: app/models/services/facebook.rb [hunk 1]
```diff
@@ -23,16 +23,16 @@ def post_to_facebook(url, body)
...
+    else
+      post_url = short_post_url(post, protocol: AppConfig.pod_uri.scheme, host: AppConfig.pod_uri.authority)
+      message = "#{post_message} (via #{post_url} )"
     end
 
-    {message: message,
-     access_token: access_token,
-     link:  post.message.urls.first
-    }
+    {message: message, access_token: access_token, link: post.message.urls.first}
```
Comment (SuperTux88) [inline app/models/services/facebook.rb]: > Align the elements of a hash literal if they span more than one line

It means to write it like this:

```ruby
    {
      message:      message,
      access_token: access_token,
      link:         post.message.urls.first
    }
```

But I think it's OK to have it in one line here.

File: app/models/services/facebook.rb [hunk 2]
```diff
@@ -23,16 +23,16 @@ def post_to_facebook(url, body)
   end
 
   def create_post_params(post)
-    message = post.message.plain_text_without_markdown
-    if post.photos.any?
-      message += " " + short_post_url(post, protocol: AppConfig.pod_uri.scheme,
-                                            host: AppConfig.pod_uri.authority)
+    post_message = post.message.plain_text_without_markdown
+
+    if post_message.empty? && post.photos.none?
```
Comment (SuperTux88) [inline app/models/services/facebook.rb]: A post should never be empty and have no photos, there is a validation for that:

https://github.com/diaspora/diaspora/blob/beeb073dbe0d95b0dd38c91f8b129fc0039bd3a0/app/models/status_message.rb#L20-L21

File: app/models/services/facebook.rb [hunk 3]
```diff
@@ -23,16 +23,16 @@ def post_to_facebook(url, body)
...
+      message = post_message
+    else
+      post_url = short_post_url(post, protocol: AppConfig.pod_uri.scheme, host: AppConfig.pod_uri.authority)
+      message = "#{post_message} (via #{post_url} )"
```
Comment (SuperTux88) [inline app/models/services/facebook.rb]: When there is no text but only photos, I think `post_message` is empty. In this case I think it looks weird to have the post be just a "(via https://pod.example.org/p/1234)", so I think it's better to just post the link in that case, as it was before. As soon as the message isn't empty, add it with (via ...).

Also, does facebook really need the space before the closing brace to detect the link? As you can see above, GitHub and also diaspora does detect URLs perfectly without this space.

File: app/models/services/facebook.rb [hunk 4]
```diff
@@ -23,16 +23,16 @@ def post_to_facebook(url, body)
   end
 
   def create_post_params(post)
-    message = post.message.plain_text_without_markdown
-    if post.photos.any?
-      message += " " + short_post_url(post, protocol: AppConfig.pod_uri.scheme,
-                                            host: AppConfig.pod_uri.authority)
+    post_message = post.message.plain_text_without_markdown
```
Comment (SuperTux88) [inline app/models/services/facebook.rb]: Why did you rename this to `post_message` but then use again `message` below? I think one variable should be enough.

File: spec/models/services/facebook_spec.rb [hunk 5]
```diff
@@ -45,7 +77,7 @@
 
       @photos.each(&:save!)
 
-      @status_message = alice.build_post(:status_message, :text => "the best pebble.")
+      @status_message = alice.build_post(:status_message, text: "the best pebble.")
```
Comment (SuperTux88) [inline spec/models/services/facebook_spec.rb]: You can move everything from this line and below to the `it "should include post url in message with photos"`, because it is only used once there. That way this code doesn't run a second time for `it "should include post url in message with photos when no text message"` where it isn't needed (because you create a message without text there).

File: spec/models/services/facebook_spec.rb [hunk 6]
```diff
@@ -22,14 +22,46 @@
...
+    it "adds URL of post to end if no link at end" do
       message = "Some text."
       post = double(message: double(plain_text_without_markdown: message, urls: []), photos: [])
       post_params = @service.create_post_params(post)
-      expect(post_params[:message]).not_to include "http"
+      expect(post_params[:message]).to include "(via http:"
     end
 
-    it 'sets facebook id on post' do
+    it "adds URL at the end with ending with valid HTTP url" do
```
Comment (SuperTux88) [inline spec/models/services/facebook_spec.rb]: I don't think we need that many tests for different urls in different places in the message, because it just doesn't make any difference.

File: spec/models/services/facebook_spec.rb [hunk 7]
```diff
@@ -22,14 +22,46 @@
       post_params = @service.create_post_params(post)
     end
 
-    it 'does not add post link when no photos' do
+    it "adds URL of post to end if no link at end" do
       message = "Some text."
       post = double(message: double(plain_text_without_markdown: message, urls: []), photos: [])
       post_params = @service.create_post_params(post)
-      expect(post_params[:message]).not_to include "http"
+      expect(post_params[:message]).to include "(via http:"
```
Comment (SuperTux88) [inline spec/models/services/facebook_spec.rb]: You should check if the via link is acutally at the end if you write in the test-text "to the end".

**Review by SuperTux88 — DISMISSED** [review]
Thanks for your pull-request, here is some feedback :)

Also, can you please squash the commits when you do style fixes?

File: app/models/services/facebook.rb (hunk 2 above)
Comment (HankG) [reply app/models/services/facebook.rb]: Oh ok, I didn't know that so wanted to cover that case.  I can adjust that along with the below suggestion.

File: app/models/services/facebook.rb (hunk 1 above)
Comment (HankG) [reply app/models/services/facebook.rb]: I tried that and when I ran style check it still complained.  I'll try it again, I much prefer it over the one line...

File: spec/models/services/facebook_spec.rb (hunk 7 above)
Comment (HankG) [reply spec/models/services/facebook_spec.rb]: for each test? ok.

File: spec/models/services/facebook_spec.rb (hunk 6 above)
Comment (HankG) [reply spec/models/services/facebook_spec.rb]: It made a difference when I was putting the "via" link before the last URL, but you are right they are superfluous now.

File: spec/models/services/facebook_spec.rb (hunk 5 above)
Comment (HankG) [reply spec/models/services/facebook_spec.rb]: ok

File: app/models/services/facebook.rb (hunk 4 above)
Comment (HankG) [reply app/models/services/facebook.rb]: Code readability since that is explicitly the post message versus the message that is being embedded into the return object.  I can use one variable throughout if that is preferred.

File: spec/models/services/facebook_spec.rb (hunk 7 above)
Comment (SuperTux88) [reply spec/models/services/facebook_spec.rb]: No, as mentioned in https://github.com/diaspora/diaspora/pull/7774#discussion_r180253983 you can remove most of the tests, so you con't need to check it in each test.

File: app/models/services/facebook.rb (hunk 3 above)
Comment (SuperTux88) [reply app/models/services/facebook.rb]: Yes, please. It's already the message that is going to be posted to facebook, you only add some more things to it before you post it.

File: app/models/services/facebook.rb [hunk 8]
```diff
@@ -24,14 +24,18 @@ def post_to_facebook(url, body)
...
+    post_url = short_post_url(post, protocol: AppConfig.pod_uri.scheme, host: AppConfig.pod_uri.authority)
+
+    message = if message.empty? && post.photos.any?
+                post_url
+              else
+                "#{message} (via #{post_url})"
+              end
+
+    {
+      message: message,
```
Comment (diaspora-code-review) [inline app/models/services/facebook.rb]: Layout/AlignHash: Align the elements of a hash literal if they span more than one line.

File: app/models/services/facebook.rb [hunk 9]
```diff
@@ -24,14 +24,18 @@ def post_to_facebook(url, body)
...
+      access_token: access_token,
+      link: post.message.urls.first
```
Comment (diaspora-code-review) [inline app/models/services/facebook.rb]: Layout/AlignHash: Align the elements of a hash literal if they span more than one line.

Comment (HankG) [reply app/models/services/facebook.rb]: I switched this to this format as requested but I can't get rid of this style error.  I'm at a loss on how to make this pass the style check except making it one line.  I'm following the example in the documentation for this Rocket/Colon style (table).  Can someone point out what I'm doing wrong here?  http://www.rubydoc.info/github/bbatsov/RuboCop/RuboCop/Cop/Layout/AlignHash

**Comment by SuperTux88** [comment]
Please don't create commit messages with `Squashed commit of the following: ...` because this information isn't important afterwards. Nobody needs to know how many commits these were before and the main commit message (the first line) should always describe what was done in this commit. Also style/review fixes should always been squashed to the commit where they are introduced (if possible, for bigger PRs this sometimes would end in complicated rebases with many conflicts, so in that case a separate fix-commit is OK). For this PR there should be only one commit with `Facebook plugin always appends (via D* post URL) to cross-posts`.

By they way: You can reorder and squash commits interactively with `git rebase -i upstream/develop` (the `fixup` keyword is useful here). You can also directly commit style-fixes with `git commit --fixup=<some commit ID to which the fixup belongs>` and then you can squash them with `git rebase -i --autosquash upstream/develop`.

**Comment by HankG** [comment]
Ugh...through the wringer again...

**Comment by HankG** [comment]
So per your request not only do you want me to squash the style changes you want me to squash **all** the changes into one enormous commit, including stuff through code reviews to the very end?

**Comment by SuperTux88** [comment]
> I think this is what you were looking for now (latest commit) but I really am no longer sure anymore...

You now also included the last commit from the `develop` branch, because you somehow rebased it, that should be removed again (probably with rebasing on the `develop` branch).

> So per your request not only do you want me to squash the style changes you want me to squash all the changes into one enormous commit, including stuff through code reviews to the very end?

Not exactly, multiple commits are OK and even required in some cases (one very big commit is hard to review), but fixups for your changes (because of reviews and or style-guide) should be done in the same commit as they are introduced (so the change you did in the beginning is already clean in this commit). So you don't need the original commit and then 5 commits with "fix style", "fix style again", "maybe this time", ... ;)

Multiple commits are good when you do multiple changes in one PR that can be cleanly splitted in two commits. Maybe you do a refactoring as driveby, then the refactoring should be done in it's own commit (and style-fixes that are needed because of the refactoring should go to this commit, and other style-fixes to the other commit). Or you add a bigger feature that can be splitted in multiple parts. But this PR is a really small fix and is fine with a single commit.

Short: the commits should be done in a way they make sense in the future, when somebody reads the git log to find where some changes came from. And they don't want to find the last commit for a line to be with the message "style fixes" were `'` in this line were changed to `"`. They want to find a commit with a meaningful message where you added the line directly with `"`.

I hope it's clearer now, sorry for the confusion :)

**Comment by HankG** [comment]
> You now also included the last commit from the develop branch, because you somehow rebased it, that should be removed again (probably with rebasing on the develop branch).

I am a bit lost on how to address this, or how I would have done this in the first place.  I did the last step of the rebase -i one more time but something tells me it didn't address your comment. 

While I've used git for many years rebasing is not something we've done any of the projects I've worked on, in fact it was often verboten to do it under normal circumstances.  I logged into the dev IRC for a more interactive experience for a little while.

File: spec/models/services/facebook_spec.rb [hunk 10]
```diff
@@ -22,39 +22,57 @@
...
     end
 
-    it 'sets facebook id on post' do
-      stub_request(:post, "https://graph.facebook.com/me/feed").
-	to_return(:status => 200, :body => '{"id": "12345"}', :headers => {})
+    it "adds (via <post URL>) when message has URLs" do
```
Comment (SuperTux88) [inline spec/models/services/facebook_spec.rb]: This test is only a duplicate of the test above, because the fact that there is an URL doesn't make a difference here. So you can just remove this and remove the "when message has no URLs " from the test above.

File: app/models/services/facebook.rb [hunk 11]
```diff
@@ -24,14 +24,18 @@ def post_to_facebook(url, body)
...
-      message += " " + short_post_url(post, protocol: AppConfig.pod_uri.scheme,
-                                            host: AppConfig.pod_uri.authority)
-    end
-
-    {message: message,
-     access_token: access_token,
-     link:  post.message.urls.first
+    post_url = short_post_url(post, protocol: AppConfig.pod_uri.scheme, host: AppConfig.pod_uri.authority)
+
+    message = if message.empty? && post.photos.any?
```
Comment (SuperTux88) [inline app/models/services/facebook.rb]: When the message is empty, there always must be photos, so you can just remove the `&& post.photos.any?` here, because it's always true if the `message.empty?` is true.

**Review by SuperTux88 — CHANGES_REQUESTED** [review]
Thanks for your corrections :+1: I was on holiday last week and didn't have the time to do a review, sorry for the delay. Your PR looks almost OK, this is mostly nitpicking.

File: spec/models/services/facebook_spec.rb (hunk 10 above)
Comment (HankG) [reply spec/models/services/facebook_spec.rb]: OK. I like having this to make sure that if someone changes this later it doesn't get hosed up with embedded URLs but I'm fine making these changes.

File: app/models/services/facebook.rb (hunk 11 above)
Comment (HankG) [reply app/models/services/facebook.rb]: Isn't that implied behavior versus explicit behavior?  We only want the post format to be solely the URL if the message is empty and there is a photo.  The way the code reads it states that directly, even if the second check is not necessary.  I'll implement it either way but it seems having more explicit code on intent here is better.

File: spec/models/services/facebook_spec.rb (hunk 10 above)
Comment (SuperTux88) [reply spec/models/services/facebook_spec.rb]: If somebody changes it later, they should add a test for the changed behavior. But if somebody changes it later so it doesn't contain the "via <post URL>" when it contains a mention (just a random example which makes no sense), then you don't have a test for it. And you shouldn't add test for every random case somebody maybe could do (which doesn't make a difference now). If it makes sense to change something in the future, then we can add a test for that later (and then this test maybe should fail and will be removed).

File: app/models/services/facebook.rb (hunk 11 above)
Comment (SuperTux88) [reply app/models/services/facebook.rb]: > We only want the post format to be solely the URL if the message is empty and there is a photo. 

Well, in fact, I only want to the post format to be solely the URL if the message is empty. Because as soon as there is no message, the "via" format doesn't make sense anymore. The fact that there are photos in the original don't make any difference. At the moment it's only valid with photos when the message is empty, but maybe that changes in the future. What if it's valid to have an empty message when you have a poll for example. Then nobody would change this line here to `message.empty? && post.photos.any? || post.poll.any?` (first because everybody forgot about it, and second because it would make any sense, because the important part for the decision is only `message.empty?`) and it would look very strange if you then have an "empty" post on facebook only containing "(via <post URL>)".

Comment (HankG) [reply app/models/services/facebook.rb]: Ah true, I had been too laser focused on thinking about it with respect to the photo thing.

**Comment by HankG** [comment]
All of these elements were passing before, and pass on my machine and 7 of the 8 CI build types (in an area that has nothing to do with my code). These automated checks all worked before my last tweak, which was to half a line in the facebook.rb code.  I'm not sure what the path forward is from here...looking for guidance...

**Comment by SuperTux88** [comment]
Thanks for updating the PR, but this comment is still open: https://github.com/diaspora/diaspora/pull/7774#discussion_r185962601

> 7 of the 8 CI build types (failing in an area that has nothing to do with my code)

Some tests fail randomly from time to time, but it's not a problem when all other tests are green and the failing test has nothing to do with your change.

**Comment by HankG** [comment]
@SuperTux88 correct. I just don't know what the desired course of action is.  Is it preferred I just re-initiate the CI process and hope another random error doesn't creep in?  What is the preferred mechanism for doing that?  

I also need to figure out why editing one line made the code coverage drop since all the testing harness is unchanged.

**Comment by HankG** [comment]
I think all comments are addressed now...

File: spec/models/services/facebook_spec.rb [hunk 12]
```diff
@@ -22,39 +22,49 @@
       post_params = @service.create_post_params(post)
     end
 
-    it 'does not add post link when no photos' do
+    it "adds (via <post URL>) when message has no URLs " do
```
Comment (SuperTux88) [inline spec/models/services/facebook_spec.rb]: > I think all comments are addressed now...

The `when message has no URLs ` can be removed here, but otherwise it's OK :+1:

Comment (HankG) [reply spec/models/services/facebook_spec.rb]: Is there a way to do that through here?  This death by a thousand cuts thing is killing me.

**Comment by SuperTux88** [comment]
> Is there a way to do that through here?

I did it myself while merging it now, so I could include it to the next release. So merged as 5a8bdc1e71628238753d73e654c4eb22d3a9852f

Thank you :cookie:
